hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/DatanodeSimulationState.java

```diff
@@ -0,0 +1,375 @@
+              .setUsed(CONTAINER_SIZE)
+              .setState(ContainerReplicaProto.State.OPEN)
+              .setBlockCommitSequenceId(1000)
+              .setOriginNodeId(datanodeDetails.getUuidString())
+              .setReplicaIndex(0)
+              .build()
+      );
+    }
+  }
+
+  public synchronized void closeContainer(Long containerID) {
+    if (containers.containsKey(containerID)) {
+      containers.put(containerID, ContainerReplicaProto.State.CLOSED);
+      for (EndpointState state : endpointStates.values()) {
+        state.icr.addReport(
+            ContainerReplicaProto.newBuilder()
+                .setContainerID(containerID)
+                .setReadCount(10_000)
+                .setWriteCount(10_000)
+                .setReadBytes(10_000_000L)
+                .setWriteBytes(5_000_000_000L)
+                .setKeyCount(10_000)
+                .setUsed(5_000_000_000L)
+                .setState(ContainerReplicaProto.State.CLOSED)
+                .setBlockCommitSequenceId(1000)
+                .setOriginNodeId(datanodeDetails.getUuidString())
+                .setReplicaIndex(0)
+                .build()
+        );
+      }
+    } else {
+      LOGGER.error("Unrecognized closeContainerCommand");
+    }
+  }
+
+  @JsonSerialize(using = DatanodeDetailsSerializer.class)
+  @JsonDeserialize(using = DatanodeDeserializer.class)
+  public synchronized DatanodeDetails getDatanodeDetails() {
+    return datanodeDetails;
+  }
+
+  public synchronized void setDatanodeDetails(
+      DatanodeDetails datanodeDetails) {
+    this.datanodeDetails = datanodeDetails;
+  }
+
+  public synchronized Set<String> getPipelines() {
+    return pipelines;
+  }
+
+  public synchronized void setPipelines(Set<String> pipelines) {
+    this.pipelines = pipelines;
+  }
+
+  public synchronized boolean isRegistered() {
+    return isRegistered;
+  }
+
+  public synchronized void setRegistered(boolean registered) {
+    isRegistered = registered;
+  }
+
+  public synchronized Map<Long, ContainerReplicaProto.State> getContainers() {
+    return containers;
+  }
+
+  public synchronized void setContainers(
+      Map<Long, ContainerReplicaProto.State> containers) {
+    this.containers = containers;
+  }
+
+  public void setReadOnly(boolean readOnly) {
+    this.readOnly = readOnly;
+  }
+
+  public long getFullContainerReportDurationMs() {
+    return fullContainerReportDurationMs;
+  }
+
+  public void setFullContainerReportDurationMs(
+      long fullContainerReportDurationMs) {
+    this.fullContainerReportDurationMs = fullContainerReportDurationMs;
+  }
+
+  public int getTargetContainersCount() {
+    return targetContainersCount;
+  }
+
+  public void setTargetContainersCount(int targetContainersCount) {
+    this.targetContainersCount = targetContainersCount;
+  }
+
+  private static class DatanodeDetailsSerializer
+      extends StdSerializer<DatanodeDetails> {
+    protected DatanodeDetailsSerializer() {
```

Comment:
Hey @duongkame ~ just curious, does the protected constructor mean it'll be restricted to be called by some subclass of DatanodeDetailsSerializer?